### PR TITLE
[ADD] Added 3d cores to neuralpredictors

### PR DESCRIPTION
**Pull request by Mvystrcilova**
Added Basic3dCore and Factorized3dCore to have cores that support 3d convolutions. In `neuralpredictros/layers/encoders/encoder3d.py` is now Encoder3d as a sample how to put together a 3d core and 2d gaussian readout. Along with the cores there is now 1d laplace in regularizers to smoothen across the temporal dimension and Bias3dLayer and Scale3dLayer.

### CODE REVIEW

**Comment by pollytur** [comment]
@Mvystrcilova Thanks a lot for your contribution!  
I have some minor comments / questions 

- I would move class Core3d(Core) from [neuralpredictors/layers/cores/conv3d.py](https://github.com/sinzlab/neuralpredictors/pull/196#diff-1d1fc960e3bf20dfa59c5deec837bca2f67b3874609d8cd2d5934c89eb6199c8) to `neuralpredictors/layers/cores/base.py` as its the base class for 3D cores.
- why `Factorized3dCore(nn.Module)` does not inherit from `Core3d`? I would inherit from it and add `self.initialize(cuda=cuda)` line at the end of the `__init__` function to have consistent and fixed initialisations. Or is there any reason for it?

**Comment by Mvystrcilova** [comment]
> @Mvystrcilova Thanks a lot for your contribution! I have some minor comments / questions
> 
> * I would move class Core3d(Core) from [neuralpredictors/layers/cores/conv3d.py](https://github.com/sinzlab/neuralpredictors/pull/196#diff-1d1fc960e3bf20dfa59c5deec837bca2f67b3874609d8cd2d5934c89eb6199c8) to `neuralpredictors/layers/cores/base.py` as its the base class for 3D cores.
> * why `Factorized3dCore(nn.Module)` does not inherit from `Core3d`? I would inherit from it and add `self.initialize(cuda=cuda)` line at the end of the `__init__` function to have consistent and fixed initialisations. Or is there any reason for it?

No reason, I just forgot about it. I changed it as you suggested now.

**Comment by pollytur** [comment]
Looks fine for me. @cblessing24 could you take a look as well? I do not have the rights to approve the PR

**Comment by pollytur** [comment]
@Mvystrcilova and why there is no group sparsity for 3D cores? (Like [here](https://github.com/sinzlab/neuralpredictors/blob/24a954dacb3226ecd60b22573f686fccd0cffce5/neuralpredictors/layers/cores/conv2d.py#L295) for 2D cores) Have you tested it and it did not worked fine or its was never implemented yet? (If it was never implemented yet - its ok, we will just add it a separate PR later if it works nicely)

**Comment by Mvystrcilova** [comment]
> @Mvystrcilova and why there is no group sparsity for 3D cores? (Like [here](https://github.com/sinzlab/neuralpredictors/blob/24a954dacb3226ecd60b22573f686fccd0cffce5/neuralpredictors/layers/cores/conv2d.py#L295) for 2D cores) Have you tested it and it did not worked fine or its was never implemented yet? (If it was never implemented yet - its ok, we will just add it a separate PR later if it works nicely)

It was never implemented.

File: neuralpredictors/layers/cores/conv3d.py [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+        for l in range(0, self.layers - 1):
+            layer = OrderedDict()
+            layer[f"conv_{l + 1}"] = nn.Conv3d(
+                self.hidden_channels[l],
+                self.hidden_channels[l + 1],
+                kernel_size=self.hidden_kernel[l],
+                dilation=spatial_dilation,
+                stride=(1, self.stride[l + 1], self.stride[l + 1]),
+                padding=(0, (self.hidden_kernel[l][1]) // 2, (self.hidden_kernel[l][2]) // 2) if self.padding else 0,
+            )
```
Comment (cblessing24) [inline neuralpredictors/layers/cores/conv3d.py:183]: This could be simplified using `itertools.pairwise`.

File: neuralpredictors/layers/cores/conv3d.py [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+            if batch_norm:
+                if independent_bn_bias:
+                    layer["norm"] = nn.BatchNorm3d(hidden_channels, momentum=momentum)
+                else:
+                    layer["norm"] = nn.BatchNorm3d(hidden_channels, momentum=momentum, affine=bias and batch_norm_scale)
+                    if bias:
+                        if not batch_norm_scale:
+                            layer["bias"] = Bias3DLayer(hidden_channels)
+                    elif batch_norm_scale:
+                        layer["scale"] = Scale3DLayer(hidden_channels)
```
Comment (cblessing24) [inline neuralpredictors/layers/cores/conv3d.py:194]: Isn't this the same code as above?

File: neuralpredictors/layers/cores/conv3d.py [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+                    layer["nonlin"] = self.nonlinearities[hidden_nonlinearities]()
+
+            self.features.add_module("layer{}".format(l + 1), nn.Sequential(layer))
+
+        self.initialize(cuda=cuda)
+
+    def forward(self, x):
+        for l, features in enumerate(self.features):
+            x = features(x)
+        return x
```
Comment (cblessing24) [inline neuralpredictors/layers/cores/conv3d.py:209]: Why are you enumerating here instead of just iterating regularly?

File: neuralpredictors/layers/cores/conv3d.py [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+    def laplace_spatial(self):
+        laplace = 0
+        for filter_index in range(self.features[0].conv.weight.shape[2]):
+            laplace += self._input_weight_regularizer(self.features[0].conv.weight[:, :, filter_index, :, :])
+        return laplace
+
+    def laplace_temporal(self):
+        laplace = 0
+        for w in range(self.features[0].conv.weight.shape[-2]):
+            for h in range(self.features[0].conv.weight.shape[-1]):
```
Comment (cblessing24) [inline neuralpredictors/layers/cores/conv3d.py:220]: You could use `itertools.product` here to combine the two loops.

File: neuralpredictors/layers/cores/conv3d.py [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+                kernel_size = self.hidden_kernel[i - 1]
+                spatial_dilation = 1
+                temporal_dilation = 1
+
+            t = math.floor(t - (kernel_size[0] - 1) * temporal_dilation)
+            h = math.floor((h - (kernel_size[1] - 1) * spatial_dilation - 1) / self.stride[i] + 1)
+            w = math.floor((w - (kernel_size[2] - 1) * spatial_dilation - 1) / self.stride[i] + 1)
+
+        output_shape = self.hidden_channels[layers - 1], t, h, w
+        print("core output shape:", output_shape)
```
Comment (cblessing24) [inline neuralpredictors/layers/cores/conv3d.py:254]: Is this left over from debugging? Generally using `print` in library code is not recommended. That is what loggers are for.

File: neuralpredictors/regularizers.py [hunk 6]
```diff
@@ -291,3 +295,27 @@ def forward(self, x, avg=False):
...
+        super(Laplace1d, self).__init__()
+        filter = laplace1d()
+        self.register_buffer("filter", torch.from_numpy(filter))
+        self.padding_size = self.filter.shape[-1] // 2 if padding is None else padding
+
+    def forward(self, x):
+        return F.conv1d(x, self.filter, bias=None, padding=self.padding_size)
+
+
+class TimeLaplaceL23d(nn.Module):
```
Comment (cblessing24) [inline neuralpredictors/regularizers.py:311]: Is this class name correct?

File: neuralpredictors/regularizers.py [hunk 7]
```diff
@@ -291,3 +295,27 @@ def forward(self, x, avg=False):
         out = out * (1 - self.gaussian2d.expand(1, 1, k1, k2).to(x.device))
 
         return agg_fn(out.pow(2)) / agg_fn(x.view(oc * ic, 1, k1, k2).pow(2))
+
+
+class Laplace1d(nn.Module):
+    def __init__(self, padding):
+        super(Laplace1d, self).__init__()
```
Comment (cblessing24) [inline neuralpredictors/regularizers.py:302]: Same here.

File: neuralpredictors/layers/encoders/encoder3d.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+class Encoder3d(nn.Module):
+    def __init__(self, core, readout, readout_nonlinearity, elu_xshift, elu_yshift):
+        super().__init__()
+        self.core = core
+        self.readout = readout
+        if readout_nonlinearity == "adaptive_elu":
+            self.nonlinearity = core.nonlinearities[readout_nonlinearity](elu_xshift, elu_yshift)
+        else:
+            self.nonlinearity = core.nonlinearities[readout_nonlinearity]()
+        self.visualization_dir = None
```
Comment (cblessing24) [inline neuralpredictors/layers/encoders/encoder3d.py:14]: What is the purpose of this attribute?

File: test/layers/cores/test_conv3d.py
Comment (cblessing24) [inline test/layers/cores/test_conv3d.py:1]: I think it is great that you were thinking about adding tests to this code but an empty file does not seem useful. :P

File: neuralpredictors/layers/cores/conv3d.py
Comment (cblessing24) [inline neuralpredictors/layers/cores/conv3d.py:1]: Some of the comments I left on the basic core also apply to the factorized core. I did not look into that one in detail because it seems to contain quite a bit of duplicated code from the basic core. It would be great if that could somehow be reduced...

**Review by cblessing24 — CHANGES_REQUESTED** [review]

File: neuralpredictors/regularizers.py (hunk 7 above)
Comment (Mvystrcilova) [reply neuralpredictors/regularizers.py:302]: Yes, why wouldn't it be the correct class name? Or do you mean the arguments in super()?

File: neuralpredictors/layers/cores/conv3d.py [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+        parameters and a BiasLayer will be added after the batch_norm layer.
+        """
+        super().__init__()
+
+        regularizer_config = (
+            dict(padding=laplace_padding, kernel=input_kernel)
+            if input_regularizer == "GaussianLaplaceL2"
+            else dict(padding=laplace_padding)
+        )
+        self._input_weight_regularizer = regularizers.__dict__[input_regularizer](**regularizer_config)
```
[the remark this answers is not in the document]
Comment (Mvystrcilova) [reply neuralpredictors/layers/cores/conv3d.py:100]: This is consistent with regularizer specification in the conv2d cores. That's why it's this way.

File: neuralpredictors/regularizers.py (hunk 7 above)
Comment (cblessing24) [reply neuralpredictors/regularizers.py:302]: Yes, I was talking about the arguments to `super`.

File: neuralpredictors/layers/cores/conv3d.py
Comment (Mvystrcilova) [reply neuralpredictors/layers/cores/conv3d.py:1]: I agree, but if possible I would do some code merging in a potential future PR and not in this one.
I tried to remove the stuff you marked from both cores when present.

File: neuralpredictors/regularizers.py (hunk 6 above)
Comment (cblessing24) [reply neuralpredictors/regularizers.py]: I was asking about the name originally because the end looked like it was a typo to me at the time but know I see that that is not the case. So nevermind.
